File: learntoprogram/DeadLock.java
/**
* Description: 多线程之死锁,表现为不同锁的嵌套，两个线程分别执行时遭遇不同的嵌套顺序
* @author fnfly2005
*/
class DeadLockDemo implements Runnable
{
    private int num = 0; 
    Object obj = new Object();
    
    public void run()
    {
        while(true)
        {
            if (num++ % 2 == 1) //实现线程执行路径的切换
            {
                synchronized(DeadLockDemo.class)
                {
                    synchronized(obj)
                    {
                        System.out.println(num);
                    }
                }
            }
            else
            {
                synchronized(obj)
                {
                    synchronized(DeadLockDemo.class)
                    {
                        System.out.println(num);
                    }
                }
            }
        }
    }
}

class DeadLock
{
    public static void main(String[] args)
    {
        DeadLockDemo d = new DeadLockDemo();
        Thread t1 = new Thread(d);
        Thread t2 = new Thread(d);
        t1.start();
        t2.start();
    }
}

File: learntoprogram/DeadLock.py
#!/usr/bin/python
#coding:utf-8
"""
Description: 多线程之死锁
"""
import threading
class DeadLock(object):
    def __init__(self):
        self.num = 0
        self.r = threading.Lock()
        self.t = threading.Lock()

    def show(self):
        while True:
            if self.num % 2 == 1:
                self.r.acquire()
                self.t.acquire()
                print self.num
                self.t.release()
                self.r.release()
            else:
                self.t.acquire()
                self.r.acquire()
                print self.num
                self.r.release()
                self.t.release()
            self.num = self.num + 1

class DeadLockDemo(threading.Thread):
    def __init__(self,DeadLock):
        threading.Thread.__init__(self)
        self.DeadLock=DeadLock

    def run(self):
        self.DeadLock.show()

if __name__ == '__main__':
   d = DeadLock()
   t1 = DeadLockDemo(d)
   t2 = DeadLockDemo(d)

   t1.start()
   t2.start()
